chainio/txmgr/geometric/geometric.go

```diff
@@ -0,0 +1,609 @@
+	GasTipMultiplier:           float64(1.25),         // add an extra 25% to the gas tip
+}
+
+func fillParamsWithDefaultValues(params *GeometricTxnManagerParams) {
+	if params.ConfirmationBlocks == 0 {
+		params.ConfirmationBlocks = defaultParams.ConfirmationBlocks
+	}
+	if params.TxnBroadcastTimeout == 0 {
+		params.TxnBroadcastTimeout = defaultParams.TxnBroadcastTimeout
+	}
+	if params.TxnConfirmationTimeout == 0 {
+		params.TxnConfirmationTimeout = defaultParams.TxnConfirmationTimeout
+	}
+	if params.MaxSendTransactionRetry == 0 {
+		params.MaxSendTransactionRetry = defaultParams.MaxSendTransactionRetry
+	}
+	if params.GetTxReceiptTickerDuration == 0 {
+		params.GetTxReceiptTickerDuration = defaultParams.GetTxReceiptTickerDuration
+	}
+	if params.FallbackGasTipCap == 0 {
+		params.FallbackGasTipCap = defaultParams.FallbackGasTipCap
+	}
+	if params.GasMultiplier == 0 {
+		params.GasMultiplier = defaultParams.GasMultiplier
+	}
+	if params.GasTipMultiplier == 0 {
+		params.GasTipMultiplier = defaultParams.GasTipMultiplier
+	}
+}
+
+func NewGeometricTxnManager(
+	ethClient ethBackend,
+	wallet wallet.Wallet,
+	logger logging.Logger,
+	metrics Metrics,
+	params GeometricTxnManagerParams,
+) *GeometricTxManager {
+	fillParamsWithDefaultValues(&params)
+	return &GeometricTxManager{
+		ethClient: ethClient,
+		wallet:    wallet,
+		logger:    logger.With("component", "GeometricTxManager"),
+		metrics:   metrics,
+		params:    params,
+	}
+}
+
+// GetNoSendTxOpts This generates a noSend TransactOpts so that we can use
+// this to generate the transaction without actually sending it
+func (m *GeometricTxManager) GetNoSendTxOpts() (*bind.TransactOpts, error) {
+	ctxWithTimeout, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	from, err := m.wallet.SenderAddress(ctxWithTimeout)
+	if err != nil {
+		return nil, utils.WrapError("failed to get sender address", err)
+	}
+	return &bind.TransactOpts{
+		From:   from,
+		NoSend: true,
+		Signer: txmgr.NoopSigner,
+	}, nil
+}
+
+func newTxnRequest(tx *types.Transaction) *txnRequest {
+	return &txnRequest{
+		tx:          tx,
+		requestedAt: time.Now(),
+		txAttempts:  make([]*transaction, 0),
+	}
+}
+
+// Send is used to sign and send a transaction to an evm chain.
+// It does gas estimation and gas price bumping to ensure the transaction gets mined,
+// but it does not do nonce management, so the tx argument must have the correct nonce already set.
+//
+// Send is blocking and safe to call concurrently, so sending multiple txs in parallel is safe.
+func (t *GeometricTxManager) Send(ctx context.Context, tx *types.Transaction) (*types.Receipt, error) {
+	return t.processTransaction(ctx, newTxnRequest(tx))
+}
+
+// processTransaction sends the transaction and runs the monitoring loop which will bump the gasPrice until the tx get
+// included.
+func (t *GeometricTxManager) processTransaction(ctx context.Context, req *txnRequest) (*types.Receipt, error) {
```

Comment:
Is this method thread safe? 
Could a transaction with higher nonce end up published before one with lower nonce?

Reply:
Yes it is. Added a test to show this: [549b045](https://github.com/Layr-Labs/eigensdk-go/pull/284/commits/549b045cde1f18ddb3f2f62a3a5302ae48664ab2)
Also added comment to explain how to send txs concurrently: [e6ab890](https://github.com/Layr-Labs/eigensdk-go/pull/284/commits/e6ab89049cbe3c08b4bbca7869afeae0c1e8ca15)

Reply:
Oh btw also had to add https://github.com/Layr-Labs/eigensdk-go/pull/284/commits/3a12fcc160284c6ad2d62dbdb1bdd811dccfcab2#diff-f3751b0e1d13e5e4c58c0ae0d936505427becc99a1c52514628bdaf32df6d370R488 to slow down the gas bumping loop, otherwise txs were getting bumped like crazy if a txs with the wrong (too-high) nonce was sent. Is that normal? Was there a better way to deal with this?